README.md
# React Geosuggest [![Build Status](https://travis-ci.org/ubilabs/react-geosuggest.svg?branch=master)](https://travis-ci.org/ubilabs/react-geosuggest)

A [React](http://facebook.github.io/react/) autosuggest for the Google Maps Places API. You can also define your own suggests as defaults. Works with [Preact](https://github.com/developit/preact), too.


## Demo

Live demo: [ubilabs.github.io/react-geosuggest](http://ubilabs.github.io/react-geosuggest/)


## Installation

As this component uses the Google Maps Places API to get suggests, you must include the Google Maps Places API in the `<head>` of your HTML:

```html
<!DOCTYPE html>
  <html>
  <head>
    …
    <script src="https://maps.googleapis.com/maps/api/js?key=YOUR_API_KEY_HERE&libraries=places"></script>
  </head>
  <body>
    …
  </body>
</html>
```

Visit the [Google Developer Console](https://console.developers.google.com) to generate your API key. The API's that you have to enable in your Google API Manager Dashboard are [Google Maps Geocoding API](https://developers.google.com/maps/documentation/geocoding/start), [Google Places API Web Service](https://developers.google.com/places/web-service/) and [Google Maps Javascript API](https://developers.google.com/maps/documentation/javascript/).

The easiest way to use geosuggest is to install it from NPM and include it in your own React build process (using [Browserify](http://browserify.org), [Webpack](http://webpack.github.io/), etc).

You can also use the standalone build by including `dist/react-geosuggest.js` in your page. If you use this, make sure you have already included React, and it is available as a global variable.

```sh
npm install react-geosuggest --save
```

## Usage

The Geosuggest works out of the box by just including it. However, you can customize the behaviour with the properties noted below.

### ES6:

```js
import Geosuggest from 'react-geosuggest';

<Geosuggest />
```

### ES5:

```js
var Geosuggest = require('react-geosuggest').default;

<Geosuggest />
```

### Properties

#### placeholder
Type: `String`
Default: `Search places`

The input field will get this placeholder text.

#### initialValue
Type: `String`
Default: `''`

An initial value for the input, when you want to prefill the suggest.

#### className
Type: `String`
Default: `''`

Add an additional class to the geosuggest container.

#### style
Type: `Object`
Default: `{
  'input': {},
  'suggests': {},
  'suggestItem': {},
  'footerItem': {}
}`

Add an additional style to `Geosuggest`.
This would support overriding/adding styles to the input suggestList, suggestItem and footerItem.

#### inputClassName
Type: `String`
Default: `''`

Add an additional class to the input.

#### disabled
Type: `Boolean`
Default: `false`

Defines whether the input is disabled.

#### location
Type: `google.maps.LatLng`
Default: `null`

To get localized suggestions, define a location to bias the suggests.

#### radius
Type: `Number`
Default: `0`

The radius defines the area around the location to use for biasing the suggests. It must be accompanied by a `location` parameter.

#### bounds
Type: [`LatLngBounds`](https://developers.google.com/maps/documentation/javascript/reference?csw=1#LatLngBounds)
Default: `null`

The bounds to use for biasing the suggests. If this is set, `location` and `radius` are ignored.

#### country
Type: `String` or `Array`
Default: `null`

Restricts predictions to the specified country (ISO 3166-1 Alpha-2 country code, case insensitive). E.g., us, br, au.

#### types
Type: `Array`
Default: `null`

The types of predictions to be returned. Four types are supported: `establishment` for businesses, `geocode` for addresses, `(regions)` for administrative regions and `(cities)` for localities. If nothing is specified, all types are returned. Consult the Google Docs for [up to date types](https://developers.google.com/maps/documentation/javascript/reference#AutocompletionRequest).

#### fixtures
Type: `Array`
Default: `[]`

An array with fixtures (defaults). Each fixture has to be an object with a `label` key in it. Optionally provide a `location`, but the Geosuggest will geocode the label if no location is provided.

You can also add a `className` key to a fixture. This class will be applied to the fixture item.

#### maxFixtures
Type: `Number`
Default: `10`

Maximum number of fixtures to render.

#### googleMaps
Type: `Object`
Default: `google.maps`

In case you want to provide your own Google Maps object, pass it in as googleMaps. The default is the global google maps object.

#### ignoreTab
Type: `Boolean`
Default: `false`

When the tab key is pressed, the `onSelect` handler is invoked. Set to true to not invoke `onSelect` on tab press.

#### queryDelay
Type: `Number`
Default: `250`

Sets the delay in milliseconds after typing before a request will be sent to find suggestions.
Specify `0` if you wish to fetch suggestions after every keystroke.

#### highlightMatch
Type: `Boolean`
Default: `true`

Highlights matched text.

#### onFocus
Type: `Function`
Default: `function() {}`

Gets triggered when the input field receives focus.

#### onBlur
Type: `Function`
Default: `function(value) {}`

Gets triggered when input field loses focus.

#### onChange
Type: `Function`
Default: `function(value) {}`

Gets triggered when input field changes the value.

#### onKeyDown
Type: `Function`
Default: `function(event) {}`

Gets triggered when input field has a key pressed down. This event is triggered before onKeyPress.

#### onKeyPress
Type: `Function`
Default: `function(event) {}`

Gets triggered when input field gets key press.

#### onSuggestSelect
Type: `Function`
Default: `function(suggest) {}`

Gets triggered when a suggest got selected. Only parameter is an object with data of the selected suggest. This data is available:

* `label` – Type `String` – The label name
* `placeId` – Type `String` – If it is a preset, equals the `label`. Else it is the Google Maps `placeID`
* `location` – Type `Object` – The location containing `lat` and `lng`
* `gmaps` – Type `Object` – *Optional!* The complete response when there was a Google Maps geocode necessary (e.g. no location provided for presets). [Check the Google Maps Reference](https://developers.google.com/maps/documentation/javascript/reference#GeocoderResult) for more information on it’s structure.

#### onActivateSuggest
Type: `Function`
Default: `function(suggest) {}`

Gets triggered when a suggest is activated in the list. Only parameter is an object with data of the selected suggest. This data is available:

* `label` – Type `String` – The label name
* `placeId` – Type `String` – If it is a preset, equals the `label`. Else it is the Google Maps `placeID`

#### onSuggestNoResults
Type: `Function`
Default: `function(userInput) {}`

Gets triggered when there are no suggest results found

#### getSuggestLabel
Type: `Function`
Default: `function(suggest) { return suggest.description; }`

Used to generate a custom label for a suggest. Only parameter is a suggest (google.maps.places.AutocompletePrediction). [Check the Google Maps Reference](https://developers.google.com/maps/documentation/javascript/reference#GeocoderResult) for more information on it’s structure.

#### renderSuggestItem
Type: `Function`
Default: `null`

Used to customize the inner html of SuggestItem and allows for controlling what properties of the suggest object you want to render. Also a convenient way to add additional styling to different rendered elements within SuggestItem.

#### skipSuggest
Type: `Function`
Default: `function(suggest) {}`

If the function returns true then the suggest will not be included in the displayed results. Only parameter is an object with data of the selected suggest. (See above)

#### autoActivateFirstSuggest
Type: `Boolean`
Default: `false`

Automatically activate the first suggestion as you type. If false, the exact term(s) in the input will be used when searching and may return a result not in the list of suggestions.

#### label
Type: `String`
Default: `null`

If the `label` and a `id` prop (see "Others") were supplied, a `<label>` tag with the passed label text will be rendered. The `<label>` element's `for` attribute will correctly point to the `id` of the `<input>` element.

#### suggestsClassName
Type: `String`
Default: `''`

Add an additional class to suggest list.

#### suggestsHiddenClassName
Type: `String`
Default: `null`

Additional `className` to toggle as the list of suggestions changes visibility.

#### suggestItemClassName
Type: `String`
Default: `''`

Add an additional class to suggest item.

#### suggestItemActiveClassName
Type: `String`,
Default: `null`

Additional `className` to add when a suggestion item is active.

#### autoComplete
Type: `String`,
Default: `off`

Autocomplete input attribute.

#### footerItem
Type: `String`
Default: `null`

Defines text of footer item which is shown at the bottom of suggest list.

#### footerItemClassName
Type: `String`,
Default: `null`

Additional `className` for footer item.

#### onFooterItemSelect
Type: `Function`
Default: `function() {}`

Gets triggered when a footer item got selected.


#### Others

All [allowed attributes for `input[type="text"]`](https://github.com/ubilabs/react-geosuggest/blob/master/src/filter-input-attributes.js#L4)
All [DOM clipboard events](https://facebook.github.io/react/docs/events.html#clipboard-events).  
All [DOM mouse events](https://facebook.github.io/react/docs/events.html#mouse-events) except for drag & drop.


### Exposed component functions
These functions are accessible by setting "ref" on the component (see example below)

#### focus()
Call `focus` to focus on the element. The suggest list will be expanded with the current suggestions.

#### blur()
Call `blur` to blur (unfocus) the element. The suggest list will be closed.

#### update(value)
It is possible to update the value of the input contained within the GeoSuggest component by calling the `update` function with a new desired `value` of the type String.

#### clear()
It is also possible to clear the value of the input contained within the GeoSuggest component by calling the `clear` function.

### Example

```jsx
import React from 'react';
import ReactDOM from 'react-dom';
import Geosuggest from 'react-geosuggest';

class App extends React.Component {
  /**
   * Render the example app
   */
  render() {
    var fixtures = [
      {label: 'Old Elbe Tunnel, Hamburg', location: {lat: 53.5459, lng: 9.966576}},
      {label: 'Reeperbahn, Hamburg', location: {lat: 53.5495629, lng: 9.9625838}},
      {label: 'Alster, Hamburg', location: {lat: 53.5610398, lng: 10.0259135}}
    ];

    return (
      <div>
        <Geosuggest
          ref={el=>this._geoSuggest=el}
          placeholder="Start typing!"
          initialValue="Hamburg"
          fixtures={fixtures}
          onSuggestSelect={this.onSuggestSelect}
          location={new google.maps.LatLng(53.558572, 9.9278215)}
          radius="20"
          footerItem="Footer item" />

        {* Buttons to trigger exposed component functions *}
        <button onClick={()=>this._geoSuggest.focus()}>Focus</button>
        <button onClick={()=>this._geoSuggest.update('New Zeland')}>Update</button>
        <button onClick={()=>this._geoSuggest.clear()}>Clear</button>
      </div>
    )
  }

  /**
   * When a suggest got selected
   * @param  {Object} suggest The suggest
   */
  onSuggestSelect(suggest) {
    console.log(suggest);
  }
});

ReactDOM.render(<App />, document.getElementById('app'));
```

## Styling

This component uses [BEM](http://csswizardry.com/2013/01/mindbemding-getting-your-head-round-bem-syntax/) for namespacing the CSS classes. So styling should be easy and without conflicts. See the [geosuggest.css](https://github.com/ubilabs/react-geosuggest/blob/master/src/geosuggest.css) for an example styling.

### Note:

The `geosuggest__suggests--hidden` class is added to hide the suggestion list. You should copy the style below into your CSS file.
```css
.geosuggest__suggests--hidden {
  max-height: 0;
  overflow: hidden;
  border-width: 0;
}
```
The above class is added whenever the suggestion list needs to be hidden. This occurs when the user selects an item from the list or when the user triggers the `blur` event on the input.

Similarly, you need to have the class `geosuggest__item--active` similar to this:
```css
.geosuggest__item--active {
  background: #267dc0;
  color: #fff;
}
```
to see what item is selected, f.ex. when using the arrow keys to navigate the suggestion list.

## Contributing

Issues and pull requests are welcome!  
Please read the guidelines in [CONTRIBUTING.md](CONTRIBUTING.md) before starting to work on a PR.

## License

See [LICENSE.md](LICENSE.md)


CHANGELOG.md
## 5.0.0 (2017-07-10)

* Footer item added

## 4.1.0 (2017-07-10)


## 4.1.0 (2017-07-10)


## 4.0.0 (2017-07-10)


## 3.0.0 (2017-07-10)


### 2.3.4 (2017-06-26)


#### Bug Fixes

* **suggest-item:** add missing key prop when rendering item ([c88167e7](https://github.com/ubilabs/react-geosuggest/commit/c88167e7dcc9559d987fb81019f134ec2ba2d957))


### 2.3.3 (2017-06-19)


#### Bug Fixes

* Use filter over find to fix IE11 compatibility ([5dfff643](https://github.com/ubilabs/react-geosuggest/commit/5dfff643a3cefd2f30f4f8a5dfa706960be32831))


### 2.3.2 (2017-06-19)


#### Bug Fixes

* **builds:** fix build errors on Windows 10 ([ab2cd25e](https://github.com/ubilabs/react-geosuggest/commit/ab2cd25e16843d307b230258dd36f84092d5b64d))


### 2.3.1 (2017-06-16)


#### Bug Fixes

* **suggest-item:**
  * highlight matches in fixtures, too ([698ee71d](https://github.com/ubilabs/react-geosuggest/commit/698ee71d74cd16570df03089c56a6dae6d2fb171))
  * correct highlight of matches not at index 0 ([ac027a1d](https://github.com/ubilabs/react-geosuggest/commit/ac027a1db66830ff4bdafe98f092986f782b3664))


## 2.3.0 (2017-06-12)


#### Bug Fixes

* check whether the label is an object to use the description value ([1a5c2f78](https://github.com/ubilabs/react-geosuggest/commit/1a5c2f784aeadf52f6a247c7790b4ce37c4d702c))
* **suggest-item:** add missing key on iterated items ([b9512063](https://github.com/ubilabs/react-geosuggest/commit/b9512063462ee286de6c629518d6645f46d31cfc))

#### Features

* Expose the onKeyDown event from input ([80053d5](https://github.com/ubilabs/react-geosuggest/commit/80053d5b6359a5051285d2f08e690f1af96553b7))


## 2.2.0 (2017-06-09)


#### Bug Fixes

* check for console before calling console.error ([b6588bce](https://github.com/ubilabs/react-geosuggest/commit/b6588bce9cd4f5b88ad000cc381176cd63579277))
* allow for shift + arrow key behavior in input box ([ 	953d341](https://github.com/ubilabs/react-geosuggest/commit/953d34116856b6575cde635215b17876f52e26d1))
* **suggest-item**:
  * Items are not visible due to scroll when navigating using arrow keys ([f471f96](https://github.com/ubilabs/react-geosuggest/commit/f471f966b8d95ac7849dbd7547e888f2361845e1))
  * add renderSuggestItem prop for custom suggest item ([f0ff0eb](https://github.com/ubilabs/react-geosuggest/commit/f0ff0eb493eb01bec961ccb9dd9b88b3778dcf5e))


#### Features

* add maxFixtures prop ([f504289](https://github.com/ubilabs/react-geosuggest/commit/f504289b39b6a042da09a013c44816177336974d))
* add highlight matched text ([1607dd3](https://github.com/ubilabs/react-geosuggest/commit/1607dd30807fa0fba7d2521a97b1b38bef59dd0c))
* add autoComplete prop ([c97ea31](https://github.com/ubilabs/react-geosuggest/commit/c97ea31da2bd5a21f494b11150272fe6e0902894))
* allow arrays for country prop ([92391c4](https://github.com/ubilabs/react-geosuggest/commit/92391c41b19f8aa911b48909b78686538cc399e3))
* use react-addons-shallow-compare ([3dcf209](https://github.com/ubilabs/react-geosuggest/commit/3dcf20980f8126f65d07371ac8770f04ebbe4065))
* use prop-types ([c76d8a0](https://github.com/ubilabs/react-geosuggest/commit/c76d8a04800951193b612a17d2756e7e1e604031))


## 2.1.0 (2017-01-11)


#### Bug Fixes

* Do not use setState to save timer instance. ([3709fdb6](https://github.com/ubilabs/react-geosuggest/commit/3709fdb6f8d7d8e7133c153398be87268b445a7a))


#### Features

* changed to use reacts internal shallowCompare func ([c66051d5](https://github.com/ubilabs/react-geosuggest/commit/c66051d53507d217630ca89d2a677c93ebce7d98))
* add blur() ([d16c61](https://github.com/ubilabs/react-geosuggest/commit/d16c6105557bcca8dabc160927a448d86f06ffbe))
*  Added ARIA attributes to whitelist ([5cb3f9a6](https://github.com/ubilabs/react-geosuggest/commit/5cb3f9a6560e032bd1188fb7d1893bf7154446ac))
* **label:** add BEM class to geosuggest label  ([fb79f34e](https://github.com/ubilabs/react-geosuggest/commit/fb79f34eceed01eeaeb94e8cfe20a4b0205e908e))
* **style**:
  * allow custom hidden list class and active item class ([57338753](https://github.com/ubilabs/react-geosuggest/commit/57338753de65fffd839e7d3285e690d3fd098505))
  * Allow adding custom classes to the suggest list and suggest item ([ed69df6c](https://github.com/ubilabs/react-geosuggest/commit/ed69df6c090772861e17dcafddd31bb1157b2611))


## 2.0.1

#### Bug Fixes

* **refactor:** changed to use reacts internal shallowCompare  

## 2.0.0 (2016-10-07)


#### Bug Fixes

* **suggest-item:** bem naming ([1ab465aa](https://github.com/ubilabs/react-geosuggest/commit/1ab465aa1ea2999f01960ee0ef99d1de5621ef83))


#### Features

* **Geosuggest:** render label if id and label text where supplied ([e2cbfc86](https://github.com/ubilabs/react-geosuggest/commit/e2cbfc862b1706be215c4d29acaa26d988806f89))
* **prop-types:** add label prop types ([0345d63b](https://github.com/ubilabs/react-geosuggest/commit/0345d63b40e157c3d5c09ad0cf7c7046ffa3095c))


### 1.25.1 (2016-09-23)


#### Bug Fixes

* **package:** Specify addon as peer dependency ([7be701c1](https://github.com/ubilabs/react-geosuggest/commit/7be701c10d3c7d1cab774f157935784ecbd02312))


#### Breaking Changes

* react-addons-shallow-compare is now a peer dependency,
meaning users must install it separately from react-geosuggest.

The peer dependency for react allows one to use either react 0.14 or
0.15. However, the dependency for react-addons-shallow-compare only
allows 0.15 versions. For anyone using react 0.14 this causes warnings
from npm saying that the peer dependency for react 0.15 (specified by
react-addons-shallow-compare) is missing. This completely breaks
workflows depending on `npm shrinkwrap`.

 ([7be701c1](https://github.com/ubilabs/react-geosuggest/commit/7be701c10d3c7d1cab774f157935784ecbd02312))


## 1.25.0 (2016-08-19)


#### Bug Fixes


src/defaults.js
/* istanbul ignore next */
/**
 * Default values
 */
export default {
  fixtures: [],
  maxFixtures: 10,
  initialValue: '',
  placeholder: 'Search places',
  disabled: false,
  className: '',
  inputClassName: '',
  location: null,
  radius: null,
  bounds: null,
  country: null,
  types: null,
  queryDelay: 250,
  googleMaps: null,
  highlightMatch: true,
  onActivateSuggest: () => {},
  onSuggestSelect: () => {},
  onSuggestNoResults: () => {},
  onFocus: () => {},
  onBlur: () => {},
  onChange: () => {},
  skipSuggest: () => {},
  getSuggestLabel: suggest => suggest.description,
  renderSuggestItem: null,
  autoActivateFirstSuggest: false,
  style: {
    'input': {},
    'suggests': {},
    'suggestItem': {},
    'footerItem': {}
  },
  ignoreTab: false,
  footerItem: null,
  onFooterItemSelect: () => {}
};


src/prop-types.js
import PropTypes from 'prop-types';

/**
 * Default values
 */
export default {
  fixtures: PropTypes.array,
  maxFixtures: PropTypes.number,
  initialValue: PropTypes.string,
  placeholder: PropTypes.string,
  disabled: PropTypes.bool,
  className: PropTypes.string,
  inputClassName: PropTypes.string,
  suggestsClassName: PropTypes.string,
  suggestsHiddenClassName: PropTypes.string,
  suggestItemClassName: PropTypes.string,
  suggestItemActiveClassName: PropTypes.string,
  location: PropTypes.object,
  radius: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  bounds: PropTypes.object,
  country: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.array
  ]),
  types: PropTypes.array,
  queryDelay: PropTypes.number,
  googleMaps: PropTypes.object,
  highlightMatch: PropTypes.bool,
  onSuggestSelect: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onKeyPress: PropTypes.func,
  skipSuggest: PropTypes.func,
  getSuggestLabel: PropTypes.func,
  renderSuggestItem: PropTypes.func,
  autoActivateFirstSuggest: PropTypes.bool,
  style: PropTypes.shape({
    input: PropTypes.object,
    suggests: PropTypes.object,
    suggestItem: PropTypes.object,
    footerItem: PropTypes.object
  }),
  ignoreTab: PropTypes.bool,
  label: PropTypes.string,
  autoComplete: PropTypes.string,
  footerItem: PropTypes.string,
  onFooterItemSelect: PropTypes.func,
  footerItemClassName: PropTypes.string
};


example/src/app.js
/* global google */

import React from 'react';
import ReactDOM from 'react-dom';
import Geosuggest from '../../src/Geosuggest';

class App extends React.Component {
  /**
   * When the input receives focus
   */
  onFocus() {
    console.log('onFocus'); // eslint-disable-line
  }

  /**
   * When the input loses focus
   * @param {String} value The user input
   */
  onBlur(value) {
    console.log('onBlur', value); // eslint-disable-line
  }

  /**
   * When the input got changed
   * @param {String} value The new value
   */
  onChange(value) {
    console.log('input changes to :' + value); // eslint-disable-line
  }

  /**
   * When a suggest got selected
   * @param  {Object} suggest The suggest
   */
  onSuggestSelect(suggest) {
    console.log(suggest); // eslint-disable-line
  }

  /**
   * When there are no suggest results
   * @param {String} userInput The user input
   */
  onSuggestNoResults(userInput) {
    console.log('onSuggestNoResults for :' + userInput); // eslint-disable-line
  }

  /**
   * When a footer item got selected
   */
  onFooterItemSelect() {
    console.log('onFooterItemSelect'); // eslint-disable-line
  }

  /**
   * Render the example app
   * @return {Function} React render function
   */
  render() {
    var fixtures = [
      {label: 'New York', location: {lat: 40.7033127, lng: -73.979681}},
      {label: 'Rio', location: {lat: -22.066452, lng: -42.9232368}},
      {label: 'Tokyo', location: {lat: 35.673343, lng: 139.710388}}
    ];

    return ( // eslint-disable-line
      <div>
        <Geosuggest
          fixtures={fixtures}
          onFocus={this.onFocus}
          onBlur={this.onBlur}
          onChange={this.onChange}
          onSuggestSelect={this.onSuggestSelect}
          onSuggestNoResults={this.onSuggestNoResults}
          location={new google.maps.LatLng(53.558572, 9.9278215)}
          radius="20"
          footerItem="Footer item"
          onFooterItemSelect={this.onFooterItemSelect}
        />
      </div>
    );
  }
}

ReactDOM.render(<App />, document.getElementById('app')); // eslint-disable-line
